compiler/rustc_resolve/src/check_unused.rs

```diff
@@ -327,7 +327,45 @@ impl Resolver<'_> {
                         || module.ends_with("_test")
                         || module.ends_with("_tests") =>
                 {
-                    Some(parent_module.span)
+                    // If the parent module has the `#[cfg(test)]` attribute, suppress a help to add
+                    // the attribute; however, if not, suggest adding the attribute. There is no way
+                    // to retrieve attributes here because we no longer have `TyCtxt`. Thus, we must
+                    // detect the attribute by manipulating the source code as a string to avoid
+                    // redundant suggestions.
+                    let sm = visitor.r.session.source_map();
+                    match sm.span_extend_to_prev_str(
+                        parent_module.span,
+                        "#[cfg(test)]",
+                        false,
+                        false,
+                    ) {
+                        Some(span) => {
+                            let cfg_test_attr_span = Span::new(
+                                span.lo() - BytePos::from_usize("#[cfg(test)]".len()),
+                                span.lo(),
+                                span.ctxt(),
+                                span.parent(),
+                            );
+
+                            // We still have to keep in mind commented attributes by line comments:
+                            //
+                            // // #[cfg(test)]
+                            //
+                            // However, we do not need to care about block comment cases so that the
+                            // call of `span_extend_to_prev_str` does not accept newlines and it
+                            // would be led to the `None => Some((parent_module.span, None))`
+                            // branch.
+                            //
+                            // /* #[cfg(test)] */
+                            if let Some(sp) = sm.span_extend_to_prev_str(cfg_test_attr_span, "//", false, false)
```

Comment:
We tend to avoid giving any semantic meaning to comments during compilation.  Comments are stripped from the internal representations for precisely this reason.